### ISSUE TITLE
Report incompatible operands on class enums

### ISSUE THREAD

**Issue by hydroper** [microsoft/TypeScript#54732]
# Suggestion

## 🔍 Search Terms

Incompatible operands

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Equality expressions where the type of both operands is known at compile-time (i.e. not `any`) and are incompatible (e.g. do not _intersect_ nor are subtypes of the other, or `any`), should report an error.

## 📃 Motivating Example

```ts
class C1 {
    static readonly X = new C1;
}
class C2 {
    static readonly X = new C2;
}
C1.X === C2.X; // should report an error since C1 and C2 are unrelated
```

## 💻 Use Cases

People emulating Java enums which are classes with custom properties and methods benefit from that error.

Say you've a `PS3Item`, which holds a `PS3ItemType`: you may accidentally compare a `PS3Item` with a `PS3ItemType`. If a `===` expression knows the type of both operands and given that:

- `PS3Item` is not compatible with `PS3ItemType` (e.g. not a subset) and
- `PS3ItemType` is not compatible with `PS3Item` (e.g. not a subset)

That error would not break program compatibility. If e.g. the `any` type is one of the operands, the error doesn't occur, including for union types which may intersect.

**Comment by MartinJohns** [comment]
FYI, it's perfectly valid that this comparison can succeed. `C1` and `C2` are structurally compatible, so they can be interchanged one for another. And `readonly` applies to this interface, but it's implicitly compatible with the mutable equivalent.

**Comment by hydroper** [comment]
@MartinJohns This seems to be the case for TypeScript. For Java this produces a compile-time error:

```java
public class Program
{
    public enum C1 {
        X;
    }
    public enum C2 {
        X;
    }
    public static void main(String[] args) {
		System.out.println(C1.X == C2.X);
	}
}
```

**Comment by MartinJohns** [comment]
Well, yeah, different languages have different semantics. TypeScript is not Java, and imitating other languages is an explicit non-goal of TypeScript. I don't understand what you're trying to say with your comment.

**Comment by fatcerberus** [comment]
Yes, Java is nominally typed.  TypeScript is structurally typed, so considers `C1` and `C2` to be *the same type*.

**Comment by fatcerberus** [comment]
Related FAQ entry: https://github.com/microsoft/TypeScript/wiki/FAQ#when-and-why-are-classes-nominal
